File: python/decode_strings.py
class Solution:
    def decodeString(self, s: str) -> str:
        """time complexity: O(n^2)"""
        stack: list[str] = list()
        for i in range(len(s)):
            if s[i] == "]":
                accu: str = ""
                while len(stack) > 0 and stack[-1] != "[":
                    accu = stack.pop() + accu
                stack.pop()
                times: str = ""
                while len(stack) > 0 and stack[-1].isdigit():
                    times = stack.pop() + times
                k: int = int(times)
                stack.append(accu * k)
            else:
                stack.append(s[i])

        return "".join(stack)

    def decodeString2(self, s: str) -> str:
        """time complexity: O(n)"""
        _, result = self.helper(0, s)
        return result

    def helper(self, i: int, s: str) -> tuple[int, str]:
        num: int = 0
        result: str = ""
        while i < len(s):
            if s[i].isdigit():
                num = num * 10 + int(s[i])
            elif s[i] == "[":
                i, subs = self.helper(i + 1, s)
                result = result + num * subs
                num = 0
            elif s[i] == "]":
                return i, result
            else:
                result += s[i]
            i += 1
        return i, result


def test_decodeString_case_1():
    # arrange
    s: str = "3[a]2[bc]"
    expected: str = "aaabcbc"

    # act
    solution = Solution()
    actual = solution.decodeString(s)

    # assert
    assert expected == actual


def test_decodeString_case_2():
    # arrange
    s: str = "3[a2[c]]"
    expected: str = "accaccacc"

    # act
    solution = Solution()
    actual = solution.decodeString(s)

    # assert
    assert expected == actual


def test_decodeString_case_3():
    # arrange
    s: str = "2[abc]3[cd]ef"
    expected: str = "abcabccdcdcdef"

    # act
    solution = Solution()
    actual = solution.decodeString(s)

    # assert
    assert expected == actual


def test_decodeString_case_4():
    # arrange
    s: str = "abc3[cd]xyz"
    expected: str = "abccdcdcdxyz"

    # act
    solution = Solution()
    actual = solution.decodeString(s)

    # assert
    assert expected == actual


def test_decodeString_case_5():
    # arrange
    s: str = "100[leetcode]"
    expected: str = (
        "leetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcode"
    )

    # act
    solution = Solution()
    actual = solution.decodeString(s)

    # assert
    assert expected == actual


def test_decodeString_case_6():
    # arrange
    s: str = "100[leetcode]"
    expected: str = (
        "leetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcode"
    )

    # act
    solution = Solution()
    actual = solution.decodeString2(s)

    # assert
    assert expected == actual


def test_decodeString_case_7():
    # arrange
    s: str = "3[a]2[bc]"
    expected: str = "aaabcbc"

    # act
    solution = Solution()
    actual = solution.decodeString2(s)

    # assert
    assert expected == actual

File: rust/src/code/decode_strings.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(n^2)
    pub fn decode_string(s: String) -> String {
        use std::collections::VecDeque;

        let mut stack: Vec<char> = Vec::new();
        let s: Vec<char> = s.chars().collect();
        for i in 0..s.len() {
            if s[i] != ']' {
                stack.push(s[i]);
            } else {
                let mut accu: VecDeque<char> = VecDeque::new();
                while stack.len() > 0 && stack[stack.len() - 1] != '[' {
                    accu.push_front(stack.pop().unwrap());
                }
                stack.pop().unwrap();
                let mut times: VecDeque<char> = VecDeque::new();
                while stack.len() > 0 && stack[stack.len() - 1].is_digit(10) {
                    times.push_front(stack.pop().unwrap());
                }
                let times: i32 = times.iter().collect::<String>().parse::<i32>().unwrap();
                let mut result: Vec<char> = Vec::new();
                for _ in 0..times {
                    result.append(&mut accu.clone().into_iter().collect::<Vec<char>>());
                }
                stack.append(&mut result);
            }
        }

        stack.iter().collect::<String>()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn daily_decode_string_case_1() {
        // arrange
        let s: String = "2[abc]3[cd]ef".to_string();
        let expected: String = "abcabccdcdcdef".to_string();

        // act
        let actual = Solution::decode_string(s);

        assert_eq!(expected, actual);
    }

    #[test]
    fn daily_decode_string_case_2() {
        // arrange
        let s: String = "abc3[cd]xyz".to_string();
        let expected: String = "abccdcdcdxyz".to_string();

        // act
        let actual = Solution::decode_string(s);

        assert_eq!(expected, actual);
    }

    #[test]
    fn daily_decode_string_case_3() {
        // arrange
        let s: String = "100[leetcode]".to_string();
        let expected: String = "leetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcodeleetcode".to_string();

        // act
        let actual = Solution::decode_string(s);

        assert_eq!(expected, actual);
    }
}
